Allow an array to be a default translation value

aprescott opened the pull request:
4.2.1 introduced a change to the way `translate`/`t` works with an option of `default: [[]]`. In 4.2.0, this would give a default value of `[]`, but in 4.2.1, it leads to a missing translation.

`default: [[]]` is again allowed for cases where a default of `[]` is needed.

This should address #19640.

rafaelfranca commented:
For me does not make sense a translation returning an array. Could you get am array form the translation key, not form the default?

aprescott committed:
Allow an array to be a default translation value.

4.2.1 introduced a change to the way `translate`/`t` works with an
option of `default: [[]]`. In 4.2.0, this would give a default value of
`[]`, but in 4.2.1, it leads to a missing translation.

`default: [[]]` is again allowed for cases where a default of `[]` is
needed.

This addresses GitHub issue 19640.

aprescott commented:
@rafaelfranca I'm not quite sure what you mean by getting an array from the translation key.

Just for more context: in the app I work on, we have existing `.yml` translation files with array values, such as

```ruby
programs:
  # ...
  available_programs:
    - Foo
    - Bar
    - Baz
```

Are you suggesting that it is wrong to try to `translate` `available_programs` into an array at all? That would be a bit strange to me, given that I18n is backed by `.yml` files and using arrays is a pretty valid use case.

For fairly complex reasons, it is possible that `available_programs` is not actually specified in a `.yml`, in which case we want to fall back to a value of `[]`, which we're trying to do with `default: [[]]`.

aprescott commented:
It might be possible to work around this by using some variant of `translate(..., default: nil) || []`.

aprescott commented:
... at least after (I think) #19421 gets included in a release.

aprescott commented:
Oh, it looks like `default: nil` doesn't quite do what I expect, actually. So I guess not.

aprescott commented:
Sadly it looks like the best I have is basically

```ruby
value = translate(key, default: "___replace_me_with_default")

if value == "___replace_me_with_default"
  []
else
  value
end
```

That's less an ideal, but I don't see any alternative — I can't get `default:` to give me back `nil` (for use with `||`) or `[]`.

matthewd commented:
This reverts a (seemingly inadvertent) behavioural change, and better maintains the i18n gem's indifference to what might appear in `:default`, so it sounds fine to me. (Without opining on the reasonableness of returning a non-string from our `translate`... e.g., the `_html` magic certainly isn't expecting that.)

Or, viewed from another angle: if we intended to flatten the default, and avoid passing an array to `I18n.translate`, this would be the wrong way/place to do it.

So please do open a PR against master.

aprescott commented:
Reopened as #19649.
